### PR TITLE
Enable building rabit on Windows

### PR DESCRIPTION
**Pull request by trivialfis**
No test is performed.  Just trying to get rid of the `rabit_empty` target.

### CODE REVIEW

**Comment by hcho3** [comment]
Let’s remove the Rabit Makefile, since XGBoost doesn’t use Makefile.

**Comment by trivialfis** [comment]
@hcho3 We have R package ..

**Comment by hcho3** [comment]
Ah so Rabit is not part of the amalgamation and it needs a Makefile. Got it.

**Comment by trivialfis** [comment]
It seems autotools is picking mingw over msvc.

**Comment by trivialfis** [comment]
Turns out I can remove the makefile.

File: rabit/include/rabit/internal/socket.h [hunk 1]
```diff
@@ -21,31 +24,92 @@
...
-                       int timeout) { return WSAPoll ( pfd, nfds, timeout ); }
 #else
+
 #include <sys/poll.h>
 using SOCKET = int;
 using sock_size_t = size_t;  // NOLINT
-const int kInvalidSocket = -1;
 #endif  // defined(_WIN32)
 
+#define IS_MINGW() defined(__MINGW32__)
```
Comment (RAMitchell) [inline rabit/include/rabit/internal/socket.h:50]: Ifdef'ing everything seems quite a messy solution. Any alternatives?

**Review by RAMitchell — COMMENTED** [review]
Let me see if I understand. This PR is removing the dummy library that used to be linked into the windows Xgboost. Macros are being used instead for features that don't work on windows. Does rabit actually work in any way on windows or is it still disabled?

File: rabit/include/rabit/internal/socket.h (hunk 1 above)
Comment (trivialfis) [reply rabit/include/rabit/internal/socket.h:50]: I don't have any other options.

Comment (RAMitchell) [reply rabit/include/rabit/internal/socket.h:50]: I can think of anything better, as you said at least it's constrained to one file.

File: rabit/include/rabit/internal/utils.h [hunk 2]
```diff
@@ -180,6 +162,19 @@ inline std::FILE *FopenCheck(const char *fname, const char *flag) {
   return fp;
 }
 }  // namespace utils
+
+// Can not use std::min on Windows with msvc due to:
```
Comment (hcho3) [inline rabit/include/rabit/internal/utils.h:166]: We could define `NOMINMAX` macro to disable `MIN` and `MAX` macros from Windows.h.

File: tests/ci_build/test_r_package.py [hunk 3]
```diff
@@ -26,22 +26,13 @@ def get_mingw_bin():
...
-            cmd = ['R.exe', 'CMD', 'INSTALL', str(os.path.curdir)]
-            env = os.environ.copy()
-            env.update({'CC': CC, 'CXX': CXX})
-            subprocess.check_call(cmd, env=env)
-        elif args.compiler == 'msvc':
-            cmd = ['R.exe', 'CMD', 'INSTALL', str(os.path.curdir)]
-            env = os.environ.copy()
-            # autotool favor mingw by default.
-            env.update({'CC': 'cl.exe', 'CXX': 'cl.exe'})
-            subprocess.check_call(cmd, env=env)
```
Comment (hcho3) [inline tests/ci_build/test_r_package.py:42]: Why are we removing MSVC from the R package test?

Comment (hcho3) [reply tests/ci_build/test_r_package.py:42]: Ah so we are requiring CMake for MSVC compiler?

File: rabit/include/rabit/internal/utils.h (hunk 2 above)
Comment (trivialfis) [reply rabit/include/rabit/internal/utils.h:166]: That's a bit weird.  Can we just stick with current approach?  That feels like an invitation to header include ordering issues.

File: tests/ci_build/test_r_package.py (hunk 3 above)
Comment (trivialfis) [reply tests/ci_build/test_r_package.py:42]: Yup, I tried to debug some errors and found that it's picking mingw disregarding the env.

File: rabit/include/rabit/internal/utils.h (hunk 2 above)
Comment (hcho3) [reply rabit/include/rabit/internal/utils.h:166]: The current approach is fine.

**Review by hcho3 — APPROVED** [review]

**Pull request merged**